assets.go

```diff
@@ -1,354 +1,134 @@
-	name string
-	url  string
-}
-
-type Loader struct {
-	resources []Resource
-	images    map[string]*Texture
-	jsons     map[string]string
-	levels    map[string]*Level
-	sounds    map[string]string
-	fonts     map[string]*truetype.Font
-}
-
-func NewLoader() *Loader {
-	return &Loader{
-		resources: make([]Resource, 1),
-		images:    make(map[string]*Texture),
-		jsons:     make(map[string]string),
-		levels:    make(map[string]*Level),
-		sounds:    make(map[string]string),
-		fonts:     make(map[string]*truetype.Font),
-	}
-}
-
-func NewResource(url string) Resource {
-	kind := path.Ext(url)
-	name := path.Base(url)
-
-	if len(kind) == 0 {
-		warning("Cannot load extensionless resource " + url)
-		return Resource{}
-	}
+// FileLoader implements support for loading and releasing file resources.
+type FileLoader interface {
+	// Load loads the given resource into memory.
+	Load(url string, data io.Reader) error
 
-	return Resource{name: name, url: url, kind: kind[1:]}
-}
+	// Unload releases the given resource from memory.
+	Unload(url string) error
 
-func (l *Loader) AddFromDir(url string, recurse bool) {
-	files, err := ioutil.ReadDir(url)
-	if err != nil {
-		log.Fatal(err)
-	}
-	for _, f := range files {
-		furl := url + "/" + f.Name()
-		if !f.IsDir() {
-			Files.Add(furl)
-		} else if recurse {
-			Files.AddFromDir(furl, recurse)
-		}
-	}
+	// Resource returns the given resource, and an error if it didn't succeed
+	Resource(url string) (Resource, error)
 }
 
-func (l *Loader) Add(urls ...string) {
-	for _, u := range urls {
-		r := NewResource(u)
-		l.resources = append(l.resources, r)
-		log.Println(r)
-	}
+// Resource represents a game resource, such as an image or a sound.
+type Resource interface {
+	// URL returns the uniform resource locator of the given resource.
+	URL() string
 }
 
-func (l *Loader) Image(name string) *Texture {
-	return l.images[name]
-}
+// Files manages global resource handling of registered file formats for game
+// assets.
+var Files = &Formats{formats: make(map[string]FileLoader)}
 
-func (l *Loader) Json(name string) string {
-	return l.jsons[name]
-}
+// Formats manages resource handling of registered file formats.
+type Formats struct {
+	// formats maps from file extensions to resource loaders.
+	formats map[string]FileLoader
 
-func (l *Loader) Level(name string) *Level {
-	return l.levels[name]
+	// root is the directory which is prepended to every resource url internally
+	root string
 }
 
-func (l *Loader) Sound(name string) ReadSeekCloser {
-	f, err := os.Open(l.sounds[name])
-	if err != nil {
-		return nil
-	}
-	return f
+// SetRoot can be used to change the default directory from `assets` to whatever you want.
+//
+// Whenever `root` does not start with the directory `assets`, you will not be able to support mobile (Android/iOS)
+// since they require you to put all resources within the `assets` directory. You can, however, use subfolders within
+// the `assets` folder, and set those as `root`.
```

Comment:
Add reference which states the requirement of mobile to place files in a directory called `assets`.